### PR TITLE
feat: Construct ProtocolState from graphql json

### PR DESCRIPTION
**Pull request by hanabi1224**
**Summary of changes**
Changes introduced in this pull request:
- Construct ProtocolState from graphql json

**Reference issue to close (if applicable)**

Closes 

**Other information and links**

### CODE REVIEW

File: base/src/blockchain_state.rs [hunk 1]
```diff
@@ -71,6 +72,47 @@ pub struct BlockchainState {
...
+                    )?,
+                    pending_coinbase_aux: PendingCoinbaseAuxHash::from_str(
+                        json["stagedLedgerPendingCoinbaseAux"]
+                            .as_str()
+                            .unwrap_or_default(),
+                    )?,
+                },
+                // FIXME: missing from graphql API
+                pending_coinbase_hash: CoinBaseHash::from_str(
+                    "2n27mUhCEctJbiZQdrk3kxYc7DVHvJVDErjXrjNs7jnP3HMLKtuN",
```
[the remark this answers is not in the document]
Comment (hanabi1224) [reply base/src/blockchain_state.rs:94]: This is not constant value, the fix to the graphql API is on the way

File: base/src/blockchain_state.rs [hunk 2]
```diff
@@ -71,6 +72,47 @@ pub struct BlockchainState {
...
+                )?,
+            },
+            // FIXME: missing from graphql API
+            genesis_ledger_hash: LedgerHash::from_str(
+                "jwNYQU34Jb9FD6ZbKnWRALZqVDKbMrjZBKWFYZwAw8ZPMgv9Ld4",
```
[the remark this answers is not in the document]
Comment (hanabi1224) [reply base/src/blockchain_state.rs:99]: Ditto

File: base/src/blockchain_state.rs [hunk 3]
```diff
@@ -71,6 +72,47 @@ pub struct BlockchainState {
...
+            )?,
+            registers: BlockchainStateRegisters {
+                ledger: LedgerHash::from_str(
+                    json["snarkedLedgerHash"].as_str().unwrap_or_default(),
```
[the remark this answers is not in the document]
Comment (willemolding) [reply base/src/blockchain_state.rs:103]: Oh yeah I see.

Alternative idea - Make this explicit and add a conversion directly from `serde_json::Value` for `LedgerHash` and all other types that use this conversion. I think you could do a blanket implementation for all hash types that implement `from_str` so you would only need to write it once.

File: base/src/consensus_state.rs [hunk 4]
```diff
@@ -38,9 +40,35 @@ impl VrfOutputTruncated {
...
+    pub fn from_base64_str(s: &str) -> anyhow::Result<Self> {
+        Ok(Self::from_str(s)?)
+    }
+
+    /// To base64 string
+    pub fn to_base64_string(&self) -> anyhow::Result<String> {
+        let h: VrfOutputTruncatedBase64Json = self.clone().into();
+        let json_string = serde_json::to_string(&h)?;
+        Ok(serde_json::from_str(&json_string)?)
+    }
```
[the remark this answers is not in the document]
Comment (willemolding) [reply base/src/consensus_state.rs:54]: Oh that is a pain. Maybe use `FromString` and `Display` for the one you think they will end up keeping and add the other one as a function on the struct

**Review by willemolding — APPROVED** [review]

File: base/src/blockchain_state.rs (hunk 3 above)
[the remark this answers is not in the document]
Comment (hanabi1224) [reply base/src/blockchain_state.rs:103]: That's a great idea!

[the remark this answers is not in the document]
Comment (hanabi1224) [reply base/src/blockchain_state.rs]: fixed

**Review by sudo-shashank — APPROVED** [review]

File: base/src/consensus_state.rs (hunk 4 above)
[the remark this answers is not in the document]
Comment (hanabi1224) [reply base/src/consensus_state.rs:54]: Hmm, hard to guess, I will leave a TODO comment here

**Pull request merged**